CSUB-504: Build script for validator to manage node

pLabarta opened the pull request:
# Add first version of TypeScript Creditcoin-CLI

This PR adds `creditcoin-cli` to the `scripts` folder. It depends on `creditcoin-js` being built and packed. More information on its own README.

---
Practical tips for PR review & merge:

- [ ] All GitHub Actions report PASS
- [ ] Newly added code/functions have unit tests
  - [ ] Coverage tools report all newly added lines as covered
  - [ ] The positive scenario is exercised
  - [ ] Negative scenarios are exercised, e.g. assert on all possible errors
  - [ ] Assert on events triggered if applicable
  - [ ] Assert on changes made to storage if applicable
- [ ] Modified behavior/functions - try to make sure above test items are covered
- [ ] Integration tests are added if applicable/needed

pLabarta committed:
add: creditcoin-cli first version

pLabarta committed:
add: creditcoin-cli first version

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

pLabarta committed:
Merge branch 'dev' into cc-cli

atodorov commented on scripts/cc-cli/README.md:
```diff
@@ -0,0 +1,84 @@
+
+## Building
+
+This tool depends on `creditcoin-js`. Make sure to pack the library using `yarn pack` in the `creditcoin-js` folder before building. It should be available as `creditcoin-js-v0.9.0.tgz`.
```
Note: version number has already changed so maybe not a good idea to hard-code it in the readme.

pLabarta replied:
Changed to `creditcoin-js-vX.Y.Z.tgz`

atodorov commented on scripts/cc-cli/README.md:
````diff
@@ -0,0 +1,84 @@
+
+```
+creditcoin-cli wizard -sf stashseed -cf controllerseed -a 1000 -u ws://localhost:9944
+```
````
IMO it would be better from usability POV for the cli tools to be bundled inside the container and for them to talk directly to localhost within the container. 

That will remove the need to explicitly specify the URL and will also remove the need for validators to build the library from source and dealing with all the version changes over time, something like 

`docker exec validator_node creditcoin-cli wizard ...`

pLabarta replied:
The container you are talking about is the one built from `https://github.com/gluwa/creditcoin/blob/dev/Dockerfile`?

atodorov replied:
Yes.

atodorov commented on scripts/cc-cli/package.json:
```diff
@@ -0,0 +1,33 @@
+        "ts-node": "^10.8.0",
+        "typescript": "^5.0.3"
+    }
+}
```
Lcoation needs to be added to dependabot config.

atodorov commented on scripts/cc-cli/package.json:
```diff
@@ -0,0 +1,33 @@
+{
```
`scripts/cc-cli` needs to be added to the `javascript-*` jobs in ci.yml, see for example https://github.com/gluwa/creditcoin/blob/dev/.github/workflows/ci.yml#L384

pLabarta committed:
fix: remove hardcoded cc version from readme

pLabarta committed:
fix: remove hardcoded cc version from readme

pLabarta committed:
add: include creditcoin-cli in dependabot.yml

pLabarta committed:
add: include creditcoin-cli in dependabot.yml

pLabarta committed:
add: scripts/cc-cli included in javascript ci jobs

pLabarta committed:
add: scripts/cc-cli included in javascript ci jobs

pLabarta committed:
fix: indentation woops

pLabarta committed:
fix: indentation woops

pLabarta committed:
fix: use creditcoin-js v0.9.5

pLabarta committed:
fix: use creditcoin-js v0.9.5

pLabarta committed:
add: yarn lock

pLabarta committed:
add: yarn lock

pLabarta committed:
add: typecheck script

pLabarta committed:
add: typecheck script

pLabarta committed:
add: lint script

pLabarta committed:
add: lint script

pLabarta committed:
fix: remove unused imports & funcs

pLabarta committed:
fix: remove unused imports & funcs

pLabarta committed:
fix: remove == comparison

pLabarta committed:
fix: remove == comparison

pLabarta committed:
fix: use template literals

pLabarta committed:
fix: use template literals

pLabarta committed:
fix: multiple declarations

pLabarta committed:
fix: multiple declarations

pLabarta committed:
fix: use const

pLabarta committed:
fix: use const

pLabarta committed:
fix: missing radix params

pLabarta committed:
fix: missing radix params

pLabarta committed:
fix: use object property shorthand

pLabarta committed:
fix: use object property shorthand

pLabarta committed:
fix: remove require, use import

pLabarta committed:
fix: remove require, use import

pLabarta committed:
fix: remove @ts-ignore line

pLabarta committed:
fix: remove @ts-ignore line

pLabarta committed:
fix: add some await, remove some async

pLabarta committed:
fix: add some await, remove some async

pLabarta committed:
add: eslintrc config file

pLabarta committed:
add: eslintrc config file

pLabarta committed:
fix: missing format script

pLabarta committed:
fix: missing format script

pLabarta committed:
fix: missing check-format script

pLabarta committed:
fix: missing check-format script

pLabarta committed:
fix: run yarn format

pLabarta committed:
fix: run yarn format

pLabarta committed:
Merge branch 'dev' into cc-cli

pLabarta committed:
add: chill command

pLabarta committed:
add: chill command

pLabarta committed:
add: cli to Docker container

pLabarta committed:
add: cli to Docker container

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

pLabarta committed:
fix: clean apt-get install & consolidate RUNs

pLabarta committed:
fix: clean apt-get install & consolidate RUNs

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

pLabarta committed:
add: chill command file

pLabarta committed:
add: chill command file

pLabarta committed:
fix: add another --no-install-recommends flag

pLabarta committed:
fix: add another --no-install-recommends flag

pLabarta committed:
fix: add yarn update checksums to Dockerfile

pLabarta committed:
fix: add yarn update checksums to Dockerfile

atodorov commented on Dockerfile:
```diff
@@ -33,14 +33,41 @@ COPY chainspecs /creditcoin-node/chainspecs
 RUN source ~/.cargo/env && cargo build --release
 
+
+FROM node:16.13.0-bullseye-slim AS cli-builder
```
Add 
```
RUN useradd --home-dir /creditcoin-node --create-home creditcoin
USER creditcoin
```

So that commands below are not running as root.

pLabarta replied:
Do we need to build the image using a non-root user? It creates a bunch of issues with the node base image and I don't think it is necessary

Checking hadolinter, for example, the last user which is used for running the image is the one that should have the least permissions but does not check users during build

atodorov replied:
> Do we need to build the image using a non-root user?

I have to reply with "Do we need to build anything as root" ? 

To answer your question though: 

We don't require root privileges in order to build anything related to Creditcoin, especially not the JavaScript tarballs. For the moment that's true everywhere across our environment (CI, containers, local development) and IMO it's a good practice not to start using root unless explicitly needed.

 If the NodeJS docker image is having issues I'd say don't use it. A very easy workaround will be to use the same Ubuntu version and just install NodeJS as you do for the final image. That also has the added benefit of having the build and the runtime use exactly the same OS & NodeJS versions. WDYT ?

pLabarta replied:
Makes sense to avoid using root when possible and this is the case. For me it conflicts with using the image provided by Node, which should be "the right tool". But I like your take on just don't using it if it conflicts with our practices and having the same OS and versions on both stages.

atodorov replied:
Since you've indicated a :+1: on the comment which points to a reworked Dockerfile I will mark this as resolved. We should see CI jobs passing once you cherry-pick the commit.

atodorov commented on Dockerfile:
```diff
@@ -33,14 +33,41 @@ COPY chainspecs /creditcoin-node/chainspecs
-RUN useradd --home-dir /creditcoin-node --create-home creditcoin
+
+COPY --from=cli-builder /creditcoin-cli/creditcoin-js/creditcoin-js-v${creditcoin_js_version}.tgz /creditcoin-cli/creditcoin-js/creditcoin-js-v${creditcoin_js_version}.tgz
+COPY --from=cli-builder /creditcoin-cli/scripts/cc-cli/creditcoin-cli-v${cli_version}.tgz /creditcoin-cli/scripts/cli/creditcoin-cli-v${cli_version}.tgz
```
COPY commands should be after `USER`, see below.

atodorov commented on Dockerfile:
```diff
@@ -33,14 +33,41 @@ COPY chainspecs /creditcoin-node/chainspecs
+    curl -fsSL https://deb.nodesource.com/setup_20.x | bash - && \
+    apt-get install -y nodejs --no-install-recommends&& \
+    apt-get clean && \
+    rm -rf /var/lib/apt/lists/*
```
You can bundle these apt commands with the previous ones 10 lines above.

atodorov commented on scripts/cc-cli/README.md:
```diff
@@ -0,0 +1,84 @@
+
+Please be aware that the command line tool you are using is still in active development. It may contain bugs, incomplete features, or unexpected behavior. Exercise caution and use it at your own risk. Feedback and bug reports are greatly appreciated to help improve the tool and ensure its stability.
+
+## Building
```
Now that commands are inside the container I think this sections should go at the end of the readme file and be used only by us.

atodorov commented on scripts/cc-cli/README.md:
```diff
@@ -0,0 +1,84 @@
+- **rotate-keys**: Rotates the node keys used for validating.
+- **set-keys**: Set new keys for the controller account.
+- **validate**: Signal intention to start validating.
+- **wizard**: Run the validator setup wizard.
```
`chill` command missing. Is there a way to have a `help` command or `--help` option which will just list all of them instead of manually tracking the commands in the README. That will become out of date very soon.

atodorov commented on scripts/cc-cli/README.md:
````diff
@@ -0,0 +1,84 @@
+
+```
+creditcoin-cli new-seed
+```
````
I think we should have examples running the commands from a git checkout (or local install) and examples running them from the container itself.

atodorov commented on scripts/cc-cli/src/utils/balance.ts:
```diff
@@ -0,0 +1,51 @@
+}
+
+export function balanceFreeIsZero(balance: Balance): boolean {
+  return !(balance.free > 0);
```
How about the obvious `balance.free === 0` ? (or < some min value).

pLabarta replied:
Removed, same as balanceIsZero

atodorov commented on scripts/cc-cli/src/utils/balance.ts:
```diff
@@ -0,0 +1,51 @@
+    !(balance.free > 0) &&
+    !(balance.reserved > 0) &&
+    !(balance.miscFrozen > 0) &&
+    !(balance.feeFrozen > 0)
```
Similar question here. Why not compare directly with value zero ?

pLabarta replied:
When doing this balance checks it seemed the type returned from the account query was not a regular number and thus === checks were not working as expected. Figured out it was only used to print a different string and removed this utilities.

atodorov commented on scripts/cc-cli/.eslintrc.js:
```diff
@@ -0,0 +1,213 @@
+  plugins: [
+    // "eslint-plugin-jsdoc",
+    // "eslint-plugin-prefer-arrow",
+    // "eslint-plugin-react",
```
If not needed, let's remove the comments.

atodorov commented on scripts/cc-cli/src/api.ts:
```diff
@@ -0,0 +1,14 @@
+// Create new API instance
+export async function newApi(url: string) {
+  if (!url) {
+    url = "ws://localhost:9944";
```
why not specify this as the default value directly in the function signature ?

atodorov commented on scripts/cc-cli/src/commands/balance.ts:
```diff
@@ -0,0 +1,37 @@
+  const cmd = new Command("balance");
+  cmd.description("Get balance of an account");
+  cmd.option(
+    "-s, --seed [mneomonic]",
```
typo 

```suggestion
    "-s, --seed [mnemonic]",
```

atodorov commented on scripts/cc-cli/src/commands/bond.ts:
```diff
@@ -0,0 +1,64 @@
+  const cmd = new Command("bond");
+  cmd.description("Bond CTC from a Stash account");
+  cmd.option("-a, --amount [amount]", "Amount to bond");
+  cmd.option("-s, --seed [seed]", "Specify mnemonic phrase to bond from");
```
For the sake of consistency

```suggestion
  cmd.option("-s, --seed [mnemonic]", "Specify mnemonic phrase to bond from");
```

atodorov commented on scripts/cc-cli/src/commands/bond.ts:
```diff
@@ -0,0 +1,64 @@
+  cmd.option("-a, --amount [amount]", "Amount to bond");
+  cmd.option("-s, --seed [seed]", "Specify mnemonic phrase to bond from");
+  cmd.option(
+    "-f, --file [file]",
```
```suggestion
    "-f, --file [file-name]",
```

atodorov commented on scripts/cc-cli/src/commands/bond.ts:
```diff
@@ -0,0 +1,64 @@
+    "-f, --file [file]",
+    "Specify file with mnemonic phrase to bond from"
+  );
+  // cmd.option('-p, --password [password]', 'Specify password to use for new account')
```
Do we need this or is it a leftover which needs to be removed ?

atodorov commented on scripts/cc-cli/src/commands/bond.ts:
```diff
@@ -0,0 +1,64 @@
+  }
+
+  // If no amount error and exit
+  if (!options.amount || parseInt(options.amount, 10) < 1) {
```
Isn't `parseInt(options.amount, 10) < 1` the same as `!parseInt(options.amount, 10))` or `parseInt(options.amount, 10) > 0)`?

pLabarta replied:
Yup, makes sense to just check for `!parseInt(options.amount, 10))` since it cannot be less than 1

atodorov commented on scripts/cc-cli/src/commands/chill.ts:
```diff
@@ -0,0 +1,34 @@
+    "Signal intention to stop validating from a Controller account"
+  );
+  cmd.option(
+    "-s, --seed [mneomonic]",
```
```suggestion
    "-s, --seed [mnemonic]",
```

atodorov commented on scripts/cc-cli/src/commands/newSeed.ts:
```diff
@@ -0,0 +1,27 @@
+export function makeNewSeedCommand() {
+  const cmd = new Command("new");
+  cmd.description("Create new mnemonic seed");
+  cmd.option("-l, --lenght [word-lenght]", "Specify the amount of words");
```
```suggestion
  cmd.option("-l, --length [word-length]", "Specify the amount of words");
```

atodorov commented on scripts/cc-cli/src/commands/receive.ts:
```diff
@@ -0,0 +1,28 @@
+  const cmd = new Command("receive");
+  cmd.description("Show account address");
+  cmd.option(
+    "-s, --seed [mneomonic]",
```
```suggestion
    "-s, --seed [mnemonic]",
```

atodorov commented on scripts/cc-cli/src/commands/receive.ts:
```diff
@@ -0,0 +1,28 @@
+    "-f, --file [file-name]",
+    "Specify file with mnemonic phrase to use for new account"
+  );
+  // cmd.option('-p, --password [password]', 'Specify password to use for new account')
```
Leftover maybe ?

atodorov commented on scripts/cc-cli/src/commands/send.ts:
```diff
@@ -0,0 +1,35 @@
+  const cmd = new Command("send");
+  cmd.description("Send CTC from an account");
+  cmd.option(
+    "-s, --seed [mneomonic]",
```
```suggestion
    "-s, --seed [mnemonic]",
```

atodorov commented on scripts/cc-cli/src/commands/setKeys.ts:
```diff
@@ -0,0 +1,33 @@
+export function makeSetKeysCommand() {
+  const cmd = new Command("set-keys");
+  cmd.description("Set session keys for a Controller account");
+  cmd.option("-s, --seed [seed]", "Specify mnemonic phrase to set keys from");
```
For consistency

```suggestion
  cmd.option("-s, --seed [mnemonic]", "Specify mnemonic phrase to set keys from");
```

atodorov commented on scripts/cc-cli/src/commands/setKeys.ts:
```diff
@@ -0,0 +1,33 @@
+  cmd.description("Set session keys for a Controller account");
+  cmd.option("-s, --seed [seed]", "Specify mnemonic phrase to set keys from");
+  cmd.option(
+    "-f, --file [file]",
```
```suggestion
    "-f, --file [file-name]",
```

atodorov commented on scripts/cc-cli/src/commands/validate.ts:
```diff
@@ -0,0 +1,41 @@
+  const cmd = new Command("validate");
+  cmd.description("Signal intention to validate from a Controller account");
+  cmd.option(
+    "-s, --seed [mneomonic]",
```
```suggestion
    "-s, --seed [mnemonic]",
```

atodorov commented on scripts/cc-cli/src/commands/validate.ts:
```diff
@@ -0,0 +1,41 @@
+    "-f, --file [file-name]",
+    "Specify file with mnemonic phrase to use for new account"
+  );
+  cmd.option("--commission [commission]", "Specify commission for validator");
```
Is that a percentage value ? Maybe we can be more explicit about what's expected.

atodorov commented on scripts/cc-cli/src/commands/wizard.ts:
```diff
@@ -0,0 +1,204 @@
+    //     Node URL: ${options.url ? options.url : "ws://localhost:9944"}
+    //     Commission: ${percentFromPerbill(commission)}
+    //     Blocked: ${blocked ? "Yes" : "No"}
+    // `);
```
Let's remove this code block if not needed.

atodorov commented on scripts/cc-cli/src/index.ts:
```diff
@@ -0,0 +1,35 @@
+
+const program = new Command();
+
+program.version("0.0.1").description("Creditcoin Staking Tool");
```
Is version required? If yes, let's try keeping it consistent b/c `index.ts` and `package.json`.

atodorov commented on scripts/cc-cli/src/utils/balance.ts:
```diff
@@ -0,0 +1,51 @@
+//     console.log("Account address:", address);
+//     console.log(balance.toHuman());
+//     console.log("Available: " + balance.free / 1000000000000000000, "CTC");
+//     console.log("Bonded: " + balance.miscFrozen / 1000000000000000000, "CTC");
```
Is this something that needs to be removed or something that needs to be fixed (b/c of the TODO comment) ?

atodorov commented on scripts/cc-cli/src/utils/bond.ts:
```diff
@@ -0,0 +1,37 @@
+) {
+  const bondTx = api.tx.staking.bond(
+    controllerAddress,
+    `${amount}000000000000000000`, // Add 18 zeros to convert to micro units TODO: Improve amount handling
```
hmmm. What of I type `123` followed by a random number of zeroes (possibly 18) as input ? 

I think we should be handling numbers as numbers instead of as strings. 

Also some basic error checking may be good here. What if I specify a negative amount (could already be handled on the BE side but still, not a reason for not checking it) ?

atodorov commented on scripts/cc-cli/src/utils/bond.ts:
```diff
@@ -0,0 +1,37 @@
+    console.log(
+      "Invalid reward destination, must be one of 'staked', 'stash', or 'controller'"
+    );
+    process.exit(1);
```
Hmm, that's an interesting way of error handling. I was expecting to throw an exception. 

What do others think ?

pLabarta replied:
Been reviewing error handling and its quite messy. Several commands and functions do it differently.

I would refactor error handling to make it consistent, making all `utils` throw errors, while `commands` do the handling (exit, print errors and instructions).

atodorov reviewed (changes requested):
Added some comments and suggestions, mostly typos and improvements. 

However the docker build job is failing and it may have something to do with the order in which the commands are specified in Dockerfile.

pLabarta committed:
move warning to bottom of README

pLabarta committed:
move warning to bottom of README

pLabarta committed:
docs: add chill command & point to --help

pLabarta committed:
docs: add chill command & point to --help

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

pLabarta committed:
remove commented eslint plugins

pLabarta committed:
remove commented eslint plugins

pLabarta committed:
fix typo

Co-authored-by: Alexander Todorov <alex.todorov@gluwa.com>

pLabarta committed:
fix typo

Co-authored-by: Alexander Todorov <alex.todorov@gluwa.com>

pLabarta committed:
change seed value name to mnemonic

Co-authored-by: Alexander Todorov <alex.todorov@gluwa.com>

pLabarta committed:
change seed value name to mnemonic

Co-authored-by: Alexander Todorov <alex.todorov@gluwa.com>

pLabarta committed:
change seed value name to mnemonic

Co-authored-by: Alexander Todorov <alex.todorov@gluwa.com>

pLabarta committed:
change seed value name to mnemonic

Co-authored-by: Alexander Todorov <alex.todorov@gluwa.com>

pLabarta committed:
change file value name to file-name

Co-authored-by: Alexander Todorov <alex.todorov@gluwa.com>

pLabarta committed:
change file value name to file-name

Co-authored-by: Alexander Todorov <alex.todorov@gluwa.com>

pLabarta committed:
change seed value name to mnemonic

Co-authored-by: Alexander Todorov <alex.todorov@gluwa.com>

pLabarta committed:
change seed value name to mnemonic

Co-authored-by: Alexander Todorov <alex.todorov@gluwa.com>

pLabarta committed:
fix typo

Co-authored-by: Alexander Todorov <alex.todorov@gluwa.com>

pLabarta committed:
fix typo

Co-authored-by: Alexander Todorov <alex.todorov@gluwa.com>

pLabarta committed:
Apply suggestions from code review (typos and value names)

Co-authored-by: Alexander Todorov <alex.todorov@gluwa.com>

pLabarta committed:
Apply suggestions from code review (typos and value names)

Co-authored-by: Alexander Todorov <alex.todorov@gluwa.com>

pLabarta committed:
rm: balanceIs funcs & outdated comment

pLabarta committed:
rm: balanceIs funcs & outdated comment

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

pLabarta committed:
remove old comments

pLabarta committed:
remove old comments

pLabarta committed:
remove hardcoded version

pLabarta committed:
remove hardcoded version

pLabarta committed:
clean newApi func

pLabarta committed:
clean newApi func

pLabarta committed:
remove commented option

pLabarta committed:
remove commented option

pLabarta committed:
fix: modify amount check in bond command

pLabarta committed:
fix: modify amount check in bond command

pLabarta committed:
add: use BigInt instead of string

pLabarta committed:
add: use BigInt instead of string

atodorov commented on scripts/cc-cli/src/utils/bond.ts:
```diff
@@ -0,0 +1,42 @@
+  api: ApiPromise
+) {
+
+  const amountInMicroUnits = BigInt(amount) * BigInt(1000000000000000000); // Multiply by to convert to micro units
```
QE note: need to check what will happen if the amount entered is an invalid number as well as if it is already in a micro-units format (e.g. already has 18 trailing zeroes). 

@pLabarta do you know what are the accepted boundary values here ?

pLabarta replied:
Been looking into this. It seems there is no minimum valid transfer amount aside from micro-units limits. So min transfer would be `1e-18`. It can also fail if the amount is less than `500 * 1e-18` and the recipient account does not reach the existential deposit *after* the transaction.

If the number is invalid because it is too big the TX will fail because the account does not have enough funds. The Wizard is checking balances before calling `bond`, the Bond command is not.

pLabarta replied:
Added check for min bond amount (1 CTC), upper bound now checks against account balance

atodorov commented on scripts/cc-cli/src/utils/perbill.ts:
```diff
@@ -0,0 +1,7 @@
+
+export function percentFromPerbill(value: number) {
+  return value / 10_000_000;
+}
```
QE note: need to inspect where these 2 functions are used and what are the accepted boundary values.

pLabarta replied:
Added some checks to them. They are only used for handling commissions.

atodorov reviewed (changes requested):
Some of the CI jobs are failing so they need a patch. What worries me is that the container build job fails. 

Added  a few more comments for myself.

atodorov commented on Dockerfile:
```diff
@@ -33,14 +33,41 @@ COPY chainspecs /creditcoin-node/chainspecs
+COPY scripts/cc-cli /creditcoin-cli/scripts/cc-cli
+# RUN apt-get update && apt-get install curl -y && \
+#     curl -fsSL https://deb.nodesource.com/setup_20.x | bash - && \
+#     apt-get install -y nodejs
```
This commented out section is perhaps a leftover and needs to be removed.

atodorov committed:
Refactor how we build the docker image and its various stages

- use Ubuntu:22.04 as the common base
- use NodeJS 16.x to match CI jobs
- default to the creditcoin user and /creditcoin-node as the home
  directory
- all copied files are chown creditcoin:creditcoin

atodorov commented on Dockerfile:
```diff
@@ -33,14 +33,41 @@ COPY chainspecs /creditcoin-node/chainspecs
+COPY --from=cli-builder /creditcoin-cli/creditcoin-js/creditcoin-js-v${creditcoin_js_version}.tgz /creditcoin-cli/creditcoin-js/creditcoin-js-v${creditcoin_js_version}.tgz
+COPY --from=cli-builder /creditcoin-cli/scripts/cc-cli/creditcoin-cli-v${cli_version}.tgz /creditcoin-cli/scripts/cli/creditcoin-cli-v${cli_version}.tgz
+RUN apt-get update && apt-get install curl -y --no-install-recommends && \
+    curl -fsSL https://deb.nodesource.com/setup_20.x | bash - && \
```
Hmmm, I see yet another discrepancy here. Tarballs are built with NodeJS v16, while runtime uses v20 while in CI jobs we use v16 as well.  All of these need to be consistent.

pLabarta committed:
remove type annotation

pLabarta committed:
add checks & expected input to percent/perbill

pLabarta committed:
add checks to percentFromPerbill

pLabarta committed:
remove unused comments

pLabarta committed:
add docker reference to README

pLabarta committed:
merge Dockerfile refactor from #1087

pLabarta committed:
yarn format

zacharyfrederick reviewed (changes requested):
Can we move the cc-cli to be a top level item in the creditcoin repo. It seems important enough that we want people to be able to find it at the project's root.

nathanwhit commented on scripts/cc-cli/README.md:
````diff
@@ -0,0 +1,103 @@
+creditcoin-cli bond -f seed.txt --amount 1000 --controller 5DJ8qkxAbSVfyvorNBKt4BwDR9hUUzH8aqofuTAMTkLZtpv9
+```
+
+### Rotates of a particular node
````
```suggestion
### Rotate session keys of a particular node
```
Or something like that

nathanwhit commented on scripts/cc-cli/src/commands/balance.ts:
```diff
@@ -0,0 +1,33 @@
+}
+
+async function balanceAction(options: OptionValues) {
+  const api = await newApi(options.url);
```
Should the `url` option be defined in `makeBalanceCommand`? Or maybe it could be added at the top level so it applies to all of the subcommands

(edit: just noticed the comment in `index.ts` so I assume it's the latter)

pLabarta replied:
I thought it should be on the top level, but it seems that Commander expects subcommands to behave as independent programs. Will do some reading, there should be a way of doing it. If not, might be extracted by making some sort of "base" command which includes common features and subcommands extend.

pLabarta replied:
So there are two ways:

a) Create a BaseCommand class and make build every subcommand from it.
b) Add the option manually after including the subcommands (by iterating over the commands and adding the options)

I'm going for the latter which seems simpler for the CLI we currently have.

nathanwhit commented on scripts/cc-cli/src/commands/newSeed.ts:
```diff
@@ -0,0 +1,27 @@
+function newSeedAction(options: OptionValues) {
+  console.log("Creating new mnemonic...");
+  const mnemonic = options.lenght
+    ? mnemonicGenerate(options.lenght)
```
```suggestion
  const mnemonic = options.length
    ? mnemonicGenerate(options.length)
```
Gotta love dynamically created properties🥲

nathanwhit commented on scripts/cc-cli/src/commands/balance.ts:
```diff
@@ -0,0 +1,33 @@
+  cmd.option(
+    "-f, --file [file-name]",
+    "Specify file with mnemonic phrase to use for new account"
+  );
```
Couldn we just pass the address instead of the private key? Or at least have that be an option, since the private key isn't strictly needed

nathanwhit commented on scripts/cc-cli/src/commands/chill.ts:
```diff
@@ -0,0 +1,34 @@
+  );
+  cmd.option(
+    "-s, --seed [mnemonic]",
+    "Specify mnemonic phrase to use for new account"
```
```suggestion
    "Specify mnemonic phrase of controller account"
```

nathanwhit commented on scripts/cc-cli/src/commands/chill.ts:
```diff
@@ -0,0 +1,34 @@
+  );
+  cmd.option(
+    "-f, --file [file-name]",
+    "Specify file with mnemonic phrase to use for new account"
```
```suggestion
    "Specify file with mnemonic phrase of controller account"
```

nathanwhit commented on scripts/cc-cli/src/commands/newSeed.ts:
```diff
@@ -0,0 +1,27 @@
+  const cmd = new Command("new");
+  cmd.description("Create new mnemonic seed");
+  cmd.option("-l, --length [word-length]", "Specify the amount of words");
+  cmd.option("-s, --save [file-name]", "Save the new stash seed to a file");
```
```suggestion
  cmd.option("-s, --save [file-name]", "Save the new seed to a file");
```

Since you could also use it to create a non-stash account

nathanwhit commented on scripts/cc-cli/src/commands/newSeed.ts:
```diff
@@ -0,0 +1,27 @@
+}
+
+function newSeedAction(options: OptionValues) {
+  console.log("Creating new mnemonic...");
```
Nitpicky, but might be good to choose whether to call it mnemonic/seed and make it consistent

nathanwhit commented on scripts/cc-cli/src/commands/receive.ts:
```diff
@@ -0,0 +1,27 @@
+  );
+  cmd.option(
+    "-f, --file [file-name]",
+    "Specify file with mnemonic phrase to use for new account"
```
```suggestion
    "Specify file with mnemonic phrase of account"
```

nathanwhit commented on scripts/cc-cli/src/commands/receive.ts:
```diff
@@ -0,0 +1,27 @@
+  cmd.description("Show account address");
+  cmd.option(
+    "-s, --seed [mnemonic]",
+    "Specify mnemonic phrase to use for new account"
```
```suggestion
    "Specify mnemonic phrase to use of account"
```

nathanwhit commented on scripts/cc-cli/src/commands/rotateKeys.ts:
```diff
@@ -0,0 +1,17 @@
+async function rotateKeysAction(options: OptionValues) {
+  const api = await newApi(options.url);
+  const newKeys = await api.rpc.author.rotateKeys();
+  console.log(newKeys.toString());
```
```suggestion
  console.log("New keys: " + newKeys.toString());
```
Maybe a bit clearer than just seeing some hex strings printed

nathanwhit commented on scripts/cc-cli/src/commands/send.ts:
```diff
@@ -0,0 +1,35 @@
+  cmd.description("Send CTC from an account");
+  cmd.option(
+    "-s, --seed [mnemonic]",
+    "Specify mnemonic phrase to use for new account"
```
```suggestion
    "Specify mnemonic phrase to use for sending CTC"
```

nathanwhit commented on scripts/cc-cli/src/commands/send.ts:
```diff
@@ -0,0 +1,35 @@
+  );
+  cmd.option(
+    "-f, --file [file-name]",
+    "Specify file with mnemonic phrase to use for new account"
```
```suggestion
    "Specify file with mnemonic phrase to use for sending CTC"
```

nathanwhit commented on scripts/cc-cli/src/commands/rotateKeys.ts:
```diff
@@ -0,0 +1,17 @@
+
+export function makeRotateKeysCommand() {
+  const cmd = new Command("rotate-keys");
+  cmd.option("-u, --url [url]", "URL for the Substrate node");
```
```suggestion
  cmd.option("-u, --url [url]", "URL for the Substrate node", "ws://localhost:9944");
```
Since often this is run against a local node (especially because by default validator nodes _only_ allow this RPC to be called from localhost, since it isn't safe to expose externally)

nathanwhit commented on scripts/cc-cli/src/commands/send.ts:
```diff
@@ -0,0 +1,35 @@
+    "-f, --file [file-name]",
+    "Specify file with mnemonic phrase to use for new account"
+  );
+  cmd.option("-a, --amount [amount]", "Amount to send");
+  cmd.option("-t, --to [to]", "Specify recipient address");
```
Might need some handling for when these options are not specified

nathanwhit commented on scripts/cc-cli/src/commands/setKeys.ts:
```diff
@@ -0,0 +1,36 @@
+import { newApi } from "../api";
+import { getSeedFromOptions, initKeyringPair } from "../utils/account";
+
+export function makeSetKeysCommand() {
```
Not necessary for this PR, but as a user it might be nice to have a rotate + set keys command instead of having to call them in sequence

nathanwhit reviewed (changes requested):
Haven't finished reviewing but wanted to get a first round of comments in

pLabarta committed:
Apply suggestions: typos & wording

Co-authored-by: Nathan Whitaker <17734409+nathanwhit@users.noreply.github.com>

pLabarta committed:
Add default node URL for rotate-keys

Co-authored-by: Nathan Whitaker <17734409+nathanwhit@users.noreply.github.com>

pLabarta committed:
add: --rotate flag to setKeys command

pLabarta committed:
add: check amount & to on send

pLabarta committed:
add: balance command now based on address

pLabarta committed:
add: make --url flag available to all subcommands

pLabarta committed:
add: utils now throw errors

pLabarta committed:
add: check bonded amount is > 1

pLabarta committed:
make 'seed phrase' wording consistent

pLabarta committed:
fix: wrong comparison operator

pLabarta committed:
add: check if account has enough funds to bond

atodorov reviewed (approved):
LGTM. I will be doing more tests on these scripts very soon, as part of testing staking dashboard and becoming a validator. However I don't need to block the PR. Let's merge as soon as we have approvals from others.

nathanwhit reviewed (approved):
LGTM

atodorov committed:
Merge branch 'dev' into cc-cli

pLabarta committed:
undo merge commit

pLabarta committed:
Merge branch 'dev' into cc-cli

pLabarta committed:
Merge branch 'cc-cli' of https://github.com/gluwa/creditcoin into cc-cli
